### PR TITLE
Add Tulsa City Council spider

### PR DESCRIPTION
**Pull request by 1992tw**
## What's this PR do?
This PR adds a new spider for Tulsa City Council meetings that scrapes data from their Granicus-based video archive platform. The spider collects meeting information including titles, dates, times, agenda links, and video recordings from both historical archives (2016-2025) and upcoming scheduled meetings. It implements robust data extraction using XPath `normalize-space()` for clean text handling and Scrapy's `response.urljoin()` for reliable URL construction.

## Why are we doing this?
This addresses the need to track Tulsa City Council meetings as part of the City Scrapers project. The spider provides comprehensive coverage of council meetings (10 years of historical data plus upcoming events) to ensure citizens have access to public meeting information. This resolves the ticket for Tulsa City Council (Batch 1, Agency ID 1642). The spider has been improved based on code review feedback to follow Scrapy best practices and includes proper filtering to capture only full Council meetings while excluding committee meetings.

## Steps to manually test
1. Ensure the project is installed:
```pipenv sync --dev```

2. Activate the virtual env and enter the pipenv shell:
```pipenv shell```

3. Run the spider:
```scrapy crawl tulok_citycouncil -O test_output.csv```

4. Monitor the output and ensure no errors are raised.
5. Inspect `test_output.csv` to ensure the data looks valid (should contain meetings from 2016-2025 plus upcoming Council meetings, excluding committee meetings).
6. Ensure all tests pass
```pytest tests/test_tulok_citycouncil.py -v```

## Are there any smells or added technical debt to note?
No technical debt. The spider follows Scrapy best practices with idiomatic methods, proper XPath text normalization, comprehensive test coverage (22/22 tests passing), and handles edge cases appropriately (In Progress meetings, missing dates, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added Tulsa City Council meeting source covering archives (2016–2025) and upcoming events; extracts titles, start/end times, City Hall location, agenda and video links.

* **Bug Fixes / Improvements**
  * Improved parsing and filtering of meeting types (regular, special, emergency); excludes non‑council/committee entries and handles missing or in‑progress event links.

* **Tests**
  * Added comprehensive tests validating counts, fields, link formats, filtering, and representative meeting details.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: city_scrapers/spiders/tulok_citycouncil.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            if meeting_rows:
+                self.logger.info(f"Found {len(meeting_rows)} City Council meetings for {year}")
+                total_meetings += len(meeting_rows)
+
+                for row in meeting_rows:
+                    meeting = self._parse_meeting_row(row, response.url)
+                    if meeting:
+                        yield meeting
+
+        self.logger.info(f"Total City Council meetings found: {total_meetings}")
```
Comment (msrezaie) [inline city_scrapers/spiders/tulok_citycouncil.py]: This redundant as the framework itself counts and logs the number of scraped items.

File: city_scrapers/spiders/tulok_citycouncil.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                description="",
+                classification=CITY_COUNCIL,
+                start=start,
+                end=None,
+                all_day=False,
+                time_notes="",
+                location={
+                    "name": "City Hall",
+                    "address": "175 E 2nd St, Tulsa, OK 74103",
+                },
```
Comment (msrezaie) [inline city_scrapers/spiders/tulok_citycouncil.py]: Nitpick: better to use a variable for location and passing that in here instead for better readability.

**Comment by msrezaie** [comment]
@coderabbitai review

**Review by msrezaie — APPROVED** [review]
Looks good, great work on building your first spider.

File: city_scrapers/spiders/tulok_citycouncil.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        Panel IDs follow the pattern: CollapsiblePanel2024X, CollapsiblePanel2023X, etc.
+        where X is a digit (varies by year).
+        """
+        # Define panel IDs for each year (2016-2025)
+        # The panel ID pattern includes the year and a digit suffix
+        # note: The "1" suffix is intentional and required. It specifically targets
+        # City Council meetings (CollapsiblePanel{year}1) while excluding other
+        # panels like committees (CollapsiblePanel{year}2, CollapsiblePanel{year}3, etc.).
+        # Using a prefix selector would incorrectly match ALL panels for the year.
+        year_panel_ids = {year: f"CollapsiblePanel{year}1" for year in range(2016, 2026)}
```
[the remark this answers is not in the document]
Comment (1992tw) [reply city_scrapers/spiders/tulok_citycouncil.py]: Good catch! @duongcao-ea . I addressed the issue. Please re-review.
